### PR TITLE
Crash on invalid loss (i.e. infinite or nan)

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
In my experience, the training loss can occasionally become NaN for a variety of reasons. Regardless of the underlying cause, once the loss is NaN, it quickly propagates through the model, causing all subsequent computations to produce NaN as well.

I don't think there is a practical scenario where continuing training after detecting a NaN loss is desirable.

This PR adds a simple check after the loss is computed. If the loss is NaN, the corresponding rank immediately raises a RuntimeException and terminates. The remaining distributed ranks will then fail naturally due to the communication timeout.

I also added the following debug information to the exception message:

```python
f"debugging -- input_dict: {input_dict}, labels: {labels}, "
f"global_valid_tokens: {global_valid_tokens}"
```

I realize this is quite verbose and may produce large log messages, but I thought it could be useful for debugging. If you would prefer to keep the logs lighter, I'm happy to remove it.

### CODE REVIEW

**Review by tianyu-l — APPROVED** [review]
makes sense to me, thank you!

**Comment by tianyu-l** [comment]
lint error should be fixed

**Comment by francesco-bertolotti** [comment]
mmmh, it does not seem to be related to my modification:

```
ERROR Could not import `OpaqueBase` from `torch._library.opaque_object` [missing-module-attribute]
  --> torchtitan/distributed/deepep/hybridep.py:26:42
   |
26 | from torch._library.opaque_object import OpaqueBase, register_opaque_type
   |                                          ^^^^^^^^^^
   |
```

Am I reading this correctly?

**Comment by tianyu-l** [comment]
seems CI is broken. I'll merge.

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -799,6 +799,15 @@ def train_step(
                 labels=labels,
                 global_valid_tokens=global_valid_tokens,
             )
+
+            if not loss.isfinite():
```
Comment (fegin) [inline torchtitan/trainer.py:803]: I think this will cause a D2H sync every microbatch, which is not a good idea. cc., @tianyu-l

Comment (tianyu-l) [reply torchtitan/trainer.py:803]: is it? Could you share a code pointer or documentation for the fact? @fegin

Comment (fegin) [reply torchtitan/trainer.py:803]: Chatted offline, but repeat here for future reference. `if torch.some_ops()` requires a `item()` or a similar API to get the Python value (CPU) of `some_ops()` so that Python can do the condition, hence D2H sync.
